applications/serial_lte_modem/src/tcpip_proxy/slm_at_tcp_proxy.c

```diff
@@ -344,6 +350,33 @@ static int do_tcp_send(const u8_t *data, int datalen)
 	}
 }
 
+static int tcp_data_save(u8_t *data, u16_t length)
+{
+	u16_t sz_avail;
```

Comment:
Zephyr provides a ring buffer implementation, have you considered using it instead of implementing it on your own? Should be less error-prone:
https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/ring_buffer.h#L346
From what I understand all we need is to declare a ring buffer and then call `ring_buf_put()`/`ring_buf_get()` when needed.